File: apps/basic-integration/sveltekit/CMSaasStarter/src/lib/server/posthog-logs.ts
import { dev } from "$app/environment"
import { env } from "$env/dynamic/public"
import { OTLPLogExporter } from "@opentelemetry/exporter-logs-otlp-http"
import { LoggerProvider, SimpleLogRecordProcessor } from "@opentelemetry/sdk-logs"

const projectToken = env.PUBLIC_POSTHOG_PROJECT_TOKEN
const host = env.PUBLIC_POSTHOG_HOST

function getPostHogLogger() {
  if (!projectToken || projectToken === "your_posthog_project_token_here") {
    if (dev) {
      throw new Error(
        "PUBLIC_POSTHOG_PROJECT_TOKEN variable required by PostHog is missing or un-configured, this causes events to be silently missed. This error stops appearing once PUBLIC_POSTHOG_PROJECT_TOKEN is configured",
      )
    }
    return null
  }

  if (!host) {
    if (dev) {
      throw new Error(
        "PUBLIC_POSTHOG_HOST variable required by PostHog is missing or un-configured, this causes events to be silently missed. This error stops appearing once PUBLIC_POSTHOG_HOST is configured",
      )
    }
    return null
  }

  const exporter = new OTLPLogExporter({
    url: new URL("/i/v1/logs", host).toString(),
    headers: {
      authorization: `Bearer ${projectToken}`,
    },
  })
  const provider = new LoggerProvider({
    processors: [new SimpleLogRecordProcessor({ exporter })],
  })

  return {
    logger: provider.getLogger("cmsassstarter.posthog"),
    provider,
  }
}

const posthogLogging = getPostHogLogger()

export async function logPostHogAction(message: string): Promise<void> {
  if (!posthogLogging) {
    return
  }

  posthogLogging.logger.emit({
    body: message,
    severityText: "INFO",
  })
  await posthogLogging.provider.forceFlush()
}

File: apps/basic-integration/fastapi/fastapi3-ai-saas/app/posthog_logs.py
"""Isolated OpenTelemetry log export for PostHog."""

import logging

from opentelemetry.exporter.otlp.proto.http._log_exporter import OTLPLogExporter
from opentelemetry.sdk._logs import LoggerProvider, LoggingHandler
from opentelemetry.sdk._logs.export import BatchLogRecordProcessor

posthog_logger = logging.getLogger("app.posthog_export")
posthog_logger.setLevel(logging.INFO)
posthog_logger.propagate = False

_logger_provider: LoggerProvider | None = None


def configure_posthog_log_exporter(host: str, project_token: str) -> LoggerProvider:
    """Configure the dedicated logger to export only integration-owned records."""
    global _logger_provider

    if _logger_provider is not None:
        return _logger_provider

    logger_provider = LoggerProvider()
    exporter = OTLPLogExporter(
        endpoint=f"{host.rstrip('/')}/i/v1/logs",
        headers={"Authorization": f"Bearer {project_token}"},
    )
    logger_provider.add_log_record_processor(BatchLogRecordProcessor(exporter))
    posthog_logger.addHandler(LoggingHandler(logger_provider=logger_provider))
    _logger_provider = logger_provider
    return logger_provider
